network/src/network.rs

```diff
@@ -303,89 +279,79 @@ impl Network {
-        self.inner_dial(
-            expected_peer_id,
+    /// Dial all protocol except feeler
+    pub fn dial_all(&self, p2p_control: &mut ServiceControl, peer_id: &PeerId, addr: Multiaddr) {
+        let ids = self.get_protocol_ids(|id| id != FEELER_PROTOCOL_ID);
+        self.dial(p2p_control, peer_id, addr, DialProtocol::Multi(ids));
+    }
+
+    /// Dial just feeler protocol
+    pub fn dial_feeler(&self, p2p_control: &mut ServiceControl, peer_id: &PeerId, addr: Multiaddr) {
+        self.dial(
+            p2p_control,
+            peer_id,
             addr,
             DialProtocol::Single(FEELER_PROTOCOL_ID),
-        )
+        );
     }
+}
 
-    pub(crate) fn inner_build(
-        config: &NetworkConfig,
-        ckb_protocols: CKBProtocols,
-    ) -> Result<(Arc<Self>, P2PService, TimerRegistry, EventReceivers), Error> {
-        config.create_dir_if_not_exists()?;
-        let local_private_key = config.fetch_private_key()?;
-        // set max score to public addresses
-        let listened_addresses: FnvHashMap<Multiaddr, u8> = config
-            .public_addresses
-            .iter()
-            .map(|addr| (addr.to_owned(), std::u8::MAX))
-            .collect();
-        let peer_store: Arc<RwLock<dyn PeerStore>> = {
-            let mut peer_store =
-                SqlitePeerStore::file(config.peer_store_path().to_string_lossy().to_string())?;
-            let bootnodes = config.bootnodes()?;
-            for (peer_id, addr) in bootnodes {
-                peer_store.add_bootnode(peer_id, addr);
-            }
-            Arc::new(RwLock::new(peer_store))
-        };
-        let reserved_peers = config
-            .reserved_peers()?
-            .iter()
-            .map(|(peer_id, _)| peer_id.to_owned())
-            .collect::<Vec<_>>();
-        let peers_registry = PeersRegistry::new(
-            Arc::clone(&peer_store),
-            config.max_inbound_peers(),
-            config.max_outbound_peers(),
-            config.reserved_only,
-            reserved_peers,
-        );
-        // register protocols
+pub struct EventHandler {}
+
+impl ServiceHandle for EventHandler {
+    fn handle_error(&mut self, _context: &mut ServiceContext, error: ServiceError) {
+        debug!(target: "network", "p2p service error: {:?}", error);
+    }
+
+    fn handle_event(&mut self, _context: &mut ServiceContext, event: ServiceEvent) {
+        debug!(target: "network", "p2p service event: {:?}", event);
+    }
+}
+
+pub struct NetworkService {
+    p2p_service: Service<EventHandler>,
+    network_state: Arc<NetworkState>,
+    // Background services
+    bg_services: Vec<Box<dyn Future<Item = (), Error = ()> + Send + 'static>>,
+}
+
+impl NetworkService {
+    pub fn new(network_state: Arc<NetworkState>, protocols: Vec<CKBProtocol>) -> NetworkService {
+        let config = &network_state.config;
+
+        // == Build special protocols
+
+        // TODO: how to deny banned node to open those protocols?
```

Comment:
P2P can provide a callback to check banned list.

Reply:
Or provide a more generalized handshake callback

Reply:
Or check banned peer in `EventHandler`?